routing/missioncontrol.go

```diff
@@ -281,7 +211,108 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+
+	log.Debugf("Mission control history cleared")
+}
+
+// getEdgeProbability is expected to return the success probability of a payment
+// from fromNode along edge.
+func (m *MissionControl) getEdgeProbability(fromNode route.Vertex,
+	edge EdgeLocator, amt lnwire.MilliSatoshi) float64 {
+
+	m.Lock()
+	defer m.Unlock()
+
+	nodeHistory, ok := m.history[fromNode]
+	if !ok {
+		return 1
+	}
+
+	return m.getEdgeProbabilityForNode(nodeHistory, edge.ChannelID, amt)
+}
+
+// getEdgeProbabilityForNode estimates the probability of successfully
+// traversing a channel based on the node history.
+func (m *MissionControl) getEdgeProbabilityForNode(nodeHistory *nodeHistory,
+	channelID uint64, amt lnwire.MilliSatoshi) float64 {
+
+	now := m.now()
+
+	// Calculate the last failure of the given edge. A node failure is
+	// considered a failure that would have affected every edge. Therefore
+	// we insert a node level failure into the history of every channel.
+	var lastFailure time.Time
+	if nodeHistory.lastFail != nil {
+		lastFailure = *nodeHistory.lastFail
+	}
+
+	// Take into account a minimum penalize amount. For balance errors, a
+	// failure may be reported with such a minimum to prevent too aggresive
+	// penalization.
+	channelHistory, ok := nodeHistory.channelLastFail[channelID]
+	if ok && channelHistory.minPenalizeAmt <= amt {
+		if channelHistory.lastFail.After(lastFailure) {
+			lastFailure = channelHistory.lastFail
+		}
+	}
+
+	timeSinceLastFailure := now.Sub(lastFailure)
+
+	// Calculate success probability.
+	exp := -timeSinceLastFailure.Hours() / m.penaltyHalfLife.Hours()
+	probability := 1 - math.Pow(2, exp)
+
+	return probability
+}
+
+// createHistoryIfNotExists returns the history for the given node. If the node
+// is yet unknown, it will create an empty history structure.
+func (m *MissionControl) createHistoryIfNotExists(vertex route.Vertex) *nodeHistory {
+	if node, ok := m.history[vertex]; ok {
+		return node
+	}
+
+	node := &nodeHistory{
+		channelLastFail: make(map[uint64]*channelHistory),
+	}
+	m.history[vertex] = node
+
+	return node
+}
+
+// reportVertexFailure reports a node level failure.
+func (m *MissionControl) reportVertexFailure(v route.Vertex) {
+	log.Debugf("Reporting vertex %v failure to Mission Control", v)
+
+	now := m.now()
+
+	m.Lock()
+	defer m.Unlock()
+
+	history := m.createHistoryIfNotExists(v)
+	history.lastFail = &now
+}
+
+// reportEdgeFailure reports a channel level failure.
+//
+// TODO(roasbeef): also add value attempted to send and capacity of channel
+func (m *MissionControl) reportEdgeFailure(failedEdge edge,
+	minPenalizeAmt lnwire.MilliSatoshi) {
+
+	log.Debugf("Reporting channel %v failure to Mission Control",
+		failedEdge.channel)
+
+	now := m.now()
+
+	m.Lock()
+	defer m.Unlock()
+
+	history := m.createHistoryIfNotExists(failedEdge.from)
+	history.channelLastFail[failedEdge.channel] = &channelHistory{
+		lastFail:       now,
+		minPenalizeAmt: minPenalizeAmt,
```

Comment:
should this compute the min of the new value and the old? do we need a way to clear these when we get a success?

Reply:
Clearing on success is planned for a follow up pr. this main requirement of this pr is no regression. Assuming that payments don't happen concurrently, the current logic works roughly the same as before. The first failure sets the minPenalizeAmt and subsequent attempts will use around the same value, so their probability will be low. When a new attempt starts with a lower value, probability 1 is returned. A new chance, just as with the previous payment session scoped mission control.